torch/_torch_docs.py

```diff
@@ -3299,9 +3307,14 @@ def merge_dicts(*dicts):
 
     >>> torch.fmod(torch.tensor([-3., -2, -1, 1, 2, 3]), 2)
     tensor([-1., -0., -1.,  1.,  0.,  1.])
-    >>> torch.fmod(torch.tensor([1, 2, 3, 4, 5]), 1.5)
+    >>> torch.fmod(torch.tensor([1, 2, 3, 4, 5]), -1.5)
     tensor([1.0000, 0.5000, 0.0000, 1.0000, 0.5000])
 
+.. seealso::
+
+        :func:`torch.fmod` uses truncated (rounded towards zero) value
```

Comment:
The "see also" section goes before the "args" section

Also, wouldn't this "see also" section be for remainder? This is the fmod documentation, right?

The style would be: "... truncates (rounds towards zero) the quotient"

So for remainder: ":func:\`torch.remainder\` rounds the quotient towards the nearest even/odd."

Note that there's no notion of rounding towards the nearest integer because it fails to disambiguate how values like .5 or 1.5 are rounded. Rounding is to the nearest even or odd value. If this choice depends on the C++ compiler we're using (or a flag with that compiler) then we should note this dependency.

Reply:
Is an additional difference how the sign of the output is determined?

Reply:
On further research, while other reliable documentations do mention this formula but taking a look at the implementation, this is how both functions are implemented:

For `torch.remainder(input, other)`:

```
# Except for Vec256 implementation
r = input % other
if ( r is not zero and (r and b are of different signs) ) 
    r += b
```

For `torch.fmod(input, other)`:

```
# For integral types
r = input % other

# For floating types including float16
r = std::fmod(input, other)
```

I think, overall, it will be good and more correct to remove formulae and just mention about the signs of outputs. Because, the implementation we have in PyTorch doesn't use `trunc` or `round`, so we shouldn't mention this in the documentation.

I'm proposing that we stick to:

```
torch.remainder: Computes the element-wise remainder of division with the remainder having the same sign as the divisor other.

torch.fmod: Computes the element-wise remainder of division with the remainder having the same sign as the dividend input.
```

(which is just how the documentation was framed before, this one is just a compact and better version of it).

What do you think? @mruberry @kshitij12345

Reply:
That might be too much of a simplified code for `remainder`, as for floating point numbers we also have the cases where `r` and/or `b` can be `inf`, `-inf` and `nan`. Are these treated in the same way?

I think it is important to specify the way we round things, both verbally in the `seealso` section and more formally (see the proposal here https://github.com/pytorch/pytorch/pull/57941#discussion_r635270293) in the main documentation. After some testing, it seems that none of the following hold
```
torch.remainder(x, y) == x - torch.round(x/y)*y
torch.fmod(x, y) == x - torch.trunc(x/y)*y 
torch.fmod(x, y) == x - torch.round(x/y)*y
torch.remainder(x, y) == x - torch.trunc(x/y)*y 
```
as playing around with the following script shows:
```python
import torch
from itertools import product

a = torch.tensor([3.0, 2.9, 2.5, 2.1, 2.0, 1.9, 1.5, 1.1, 1.0, 0.9, .1])
for x, neg_x in product(a, [True, False]):
    for y, neg_y in product(a, [True, False]):
        if neg_x:
            x = -x
        if neg_y:
            y = -y
        d1 = torch.fmod(x, y)
        d2 = x - torch.trunc(x / y)*y
        if torch.dist(d1, d2) > 1e-4:
            print(f"{x} % {y} = {d1} or {d2}")
```
as such, we need to first understand better what this operation really does.

On a different note, it would be of interest to add a battery of tests for these functions and make sure that for every element in
```
xs = [float("inf"), 2, 1.5, 1.1, 1.0, 0.9, 0., -0., -0.9, -1.0, -1.1, -1.5, -2, -float("inf"), float("nan")]
for a, b in product(xs, xs):
   ....
```
and make sure that changing the dtypes of `a` and `b` from a float type to another float type does not change the result of these operations. This, of course, should be done in a different PR.

Reply:
Per offline discussion, here is the summary:

* For `fmod`:

From the implementation (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L784, it's the same for CUDA), this is how `fmod` is implemented:

`output = x % y if integral type else std::fmod(x, y)`

While this formula: `x - trunc(x/y) * y` doesn't hold for the implementation, as mentioned above by @Lezcano, and the following script confirms that:

```python
import torch
from itertools import product
a = torch.tensor([3.0, 2.9, 2.5, 2.1, 2.0, 1.9, 1.5, 1.1, 1.0, 0.9, .1])
for x, neg_x in product(a, [True, False]):
    for y, neg_y in product(a, [True, False]):
        if neg_x:
            x = -x
        if neg_y:
            y = -y
        d1 = torch.fmod(x, y)
        d2 = x - torch.trunc(x/y)*y
        if torch.dist(d1, d2) > 1e-4:
            print(f"{x} % {y} = {d1} or {d2}")
```

(we ideally expect this script not to print anything if the formulation is correct but it does)

We also think that we don't need to add too much information into the documentation, and just add what captures what the function really does (and put in non-ambiguous words). For `fmod`, we are thinking something like (as suggested by @Lezcano):

> More specifically, this function returns `input - n * output` where n is `input / output` with its fractional part truncated.

This is highly inspired from the documentation of `std::fmod` (please see: https://en.cppreference.com/w/cpp/numeric/math/fmod).

* for `remainder`:

From the implementation (see: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L213):

```python
if integral type:
    output = x % y
else if floating type:
    output = std::fmod(x, y)
    if output != 0 and ( (output < 0) != (y < 0) ):
        output += y
end if
```
The above is tested for floating types using the following script:

```python
import torch
from itertools import product
a = torch.tensor([3.0, 2.9, 2.5, 2.1, 2.0, 1.9, 1.5, 1.1, 1.0, 0.9, .1])
for x, neg_x in product(a, [True, False]):
    for y, neg_y in product(a, [True, False]):
        if neg_x:
            x = -x
        if neg_y:
            y = -y
        d1 = torch.remainder(x, y)
        d2 = torch.fmod(x, y)
        if (d2 != 0 and ((d2 < 0) != (y < 0))):
            d2 += y
        if torch.dist(d1, d2) > 1e-4:
            print(f"{x} % {y} = {d1} or {d2}")
```

(doesn't print anything - which is what we expect)

For `remainder`, here is what I think might be suitable:

> More specifically, this function returns `input - n * output` where n is `input / output` rounded towards the nearest integer (break ties with the nearest even integer).

For the difference between the two functions, here is what I'm thinking in the `seealso` section:

> `torch.fmod()` truncates (rounded towards zero) the quotient with the `output` having same sign as the dividend `input` while `torch.remainder()` rounds (towards the nearest even integer) the quotient with the output having same sign as the divisor `other`.

*References*:

1. GNU C library's documentation on `fmod` and `remainder`: https://www.gnu.org/software/libc/manual/html_node/Remainder-Functions.html
2. (May not be totally relevant here) `remainder` implementation in fdlibm: http://www.netlib.org/fdlibm/e_remainder.c, `fmod` implementation in fdlibm: http://www.netlib.org/fdlibm/e_fmod.c

Will appreciate your inputs on this, @mruberry @Lezcano @kshitij12345.

Reply:
```
More specifically, this function returns `input - n * output` where n is `input / output` with its fractional part truncated.
```
Shouldn't `ouput` -> `other`

Ref:
> The floating-point remainder of the division operation x/y calculated by this function is exactly the value x - n*y, where n is x/y with its fractional part truncated.

Also, I personally think that the documentation is too verbose (three times stating the same thing in different way).
For `fmod`, why not just add a reference to `std::fmod` something similar to 
https://github.com/pytorch/pytorch/blob/70506ffede498a3e9a8e828ca5f2b16292ebadc3/torch/_torch_docs.py#L5367

Also if the formula `out = input - torch.trunc(input / other) * other` doesn't hold, then should we mention it?

However let's wait for inputs from @Lezcano and @mruberry

Reply:
I agree with @kshitij12345 that we're making this too complicated.

For fmod: "Applies C++'s std::fmod for floating point tensors, and the modulus operation for integer tensors."

We can further add: "The result has the same sign as the dividend and its absolute value is less than the divisor's."

And add a link to the cppreference documentation for fmod so people can click "std::fmod".

Remainder is weird because it doesn't actually use the std::remainder function (https://en.cppreference.com/w/cpp/numeric/math/remainder). What's even crazier is that torch.remainder is super popular -- who knew!? @krshrimali would you file a follow-up issue to see if we can use std::remainder in the torch.remainder kernels? Are they even computing the same operation?

But for the docs I think we can just do: "Like torch.fmod (link) this applies C++'s std::fmod for floating point tensors and the modulus and the modulus operation for integer tensors. Unlike torch.fmod, however, if the sign of the modulus is different than the sign of the divisor then the divisor is added to the modulus."

Which isn't perfect but an improvement over today. We can revisit more perfect documentation if we decide to pursue the torch.remainder using std::remainder issue.

What are your thoughts?